src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/StorePal.Windows.Export.cs

```diff
@@ -102,6 +90,165 @@ public void MoveTo(X509Certificate2Collection collection)
+        }
+
+        public byte[] ExportPkcs12(PbeParameters exportParameters, SafePasswordHandle password)
+        {
+            byte[] exported = ExportPkcs12(preferAes256Sha256: true, password, out _);
+            return ReEncryptAndSealPkcs12(exported, password, exportParameters);
+        }
+
+        private unsafe byte[] ExportPkcs12(bool preferAes256Sha256, SafePasswordHandle password, out bool aes256Sha256)
+        {
+            Interop.Crypt32.DATA_BLOB dataBlob = new Interop.Crypt32.DATA_BLOB(IntPtr.Zero, 0);
+            Interop.Crypt32.PFXExportFlags flags =
+                Interop.Crypt32.PFXExportFlags.EXPORT_PRIVATE_KEYS |
+                Interop.Crypt32.PFXExportFlags.REPORT_NOT_ABLE_TO_EXPORT_PRIVATE_KEY;
+
+            Interop.Crypt32.PKCS12_PBES2_EXPORT_PARAMS* exportParams = null;
+            aes256Sha256 = preferAes256Sha256 && s_supportsAes256Sha256;
+
+            if (aes256Sha256)
+            {
+                flags |= Interop.Crypt32.PFXExportFlags.PKCS12_EXPORT_PBES2_PARAMS;
+                // PKCS12_PBES2_ALG_AES256_SHA256
+                char* algStr = stackalloc char[] { 'A', 'E', 'S', '2', '5', '6', '-', 'S', 'H', 'A', '2', '5', '6', '\0' };
+                Interop.Crypt32.PKCS12_PBES2_EXPORT_PARAMS p = new()
+                {
+                    dwSize = (uint)Marshal.SizeOf<Interop.Crypt32.PKCS12_PBES2_EXPORT_PARAMS>(),
+                    hNcryptDescriptor = 0,
+                    pwszPbes2Alg = algStr,
+                };
+                exportParams = &p;
+            }
+
+            if (!Interop.Crypt32.PFXExportCertStoreEx(_certStore, ref dataBlob, password, exportParams, flags))
+            {
+                throw Marshal.GetHRForLastWin32Error().ToCryptographicException();
+            }
+
+            byte[] pbEncoded = new byte[dataBlob.cbData];
+
+            fixed (byte* ppbEncoded = pbEncoded)
+            {
+                dataBlob.pbData = new IntPtr(ppbEncoded);
+
+                if (!Interop.Crypt32.PFXExportCertStoreEx(_certStore, ref dataBlob, password, exportParams, flags))
+                {
+                    throw Marshal.GetHRForLastWin32Error().ToCryptographicException();
+                }
+            }
+
+            return pbEncoded;
+        }
+
+        private static byte[] ReEncryptAndSealPkcs12(byte[] pkcs12, SafePasswordHandle password, PbeParameters newPbeParameters)
+        {
+            bool addedRef = false;
+
+            try
+            {
+                password.DangerousAddRef(ref addedRef);
+                ReadOnlySpan<char> passwordChars = password.DangerousGetSpan();
+                Pkcs12Info info = Pkcs12Info.Decode(pkcs12, out _, skipCopy: true);
+
+                if (!info.VerifyMac(passwordChars))
+                {
+                    Debug.Fail("Mac validation failed.");
+                    throw new CryptographicException();
+                }
```

Comment:
Not really trying to "break into jail" here... but I wonder if this would happen with a string with embedded NUL characters.

Reply:
Oh, hm. On one hand, "Don't do that". We could say "`ExportPkcs12` does not support embedded \\0s" and throw an arg validation up-front.

Reply:
I'm cool with throwing due to embedded NULs.  Shouldn't ever happen, but if it does, someone's probably happier that we blocked it than that they have a file that is hard to use on C based systems (PFXImportCertStore/OSSL)